include/boost/algorithm/apply_permutation.hpp

```diff
@@ -0,0 +1,121 @@
+/*
+  Copyright (c) Alexander Zaitsev <zamazan4ik@gmail.com>, 2017
+
+  Distributed under the Boost Software License, Version 1.0. (See
+  accompanying file LICENSE_1_0.txt or copy at
+  http://www.boost.org/LICENSE_1_0.txt)
+
+  See http://www.boost.org/ for latest version.
+
+
+  Based on https://blogs.msdn.microsoft.com/oldnewthing/20170104-00/?p=95115
+*/
+
+/// \file  apply_permutation.hpp
+/// \brief Apply permutation to a sequence.
+/// \author Alexander Zaitsev
+
+#ifndef BOOST_ALGORITHM_APPLY_PERMUTATION_HPP
+#define BOOST_ALGORITHM_APPLY_PERMUTATION_HPP
+
+#include <algorithm>
+#include <utility>
+#include <iterator>
+
+#include <boost/range/begin.hpp>
+#include <boost/range/end.hpp>
+
+namespace boost { namespace algorithm
+{
+
+/// \fn apply_permutation ( RandomAccessIterator1 item_begin, RandomAccessIterator1 item_end, RandomAccessIterator2 ind_begin )
+/// \brief Reorder item sequence with index sequence order
+///
+/// \param item_begin    The start of the item sequence
+/// \param item_end		 One past the end of the item sequence
+/// \param ind_begin     The start of the index sequence.
+///
+/// \note Item sequence size should be equal to index size. Otherwise behavior is undefined.
+///       Complexity: O(N).
+template<typename RandomAccessIterator1, typename RandomAccessIterator2>
+void
+apply_permutation(RandomAccessIterator1 item_begin, RandomAccessIterator1 item_end,
+                  RandomAccessIterator2 ind_begin)
+{
+    typedef typename std::iterator_traits<RandomAccessIterator1>::difference_type Diff;
+    Diff size = std::distance(item_begin, item_end);
+    for (Diff i = 0; i < size; ++i)
+    {
+        Diff current = i;
+        while (i != ind_begin[current])
+        {
+            Diff next = static_cast<Diff>(ind_begin[current]);
+            std::swap(item_begin[current], item_begin[next]);
```

Comment:
Rather than `std::swap()`, you should bring swap into scope with a using decl, and then call it unadorned - thereby enabling ADL. Reference here: https://stackoverflow.com/questions/28130671/how-does-using-stdswap-enable-adl

Reply:
Fixed.